[WIRES] Allow user to define non-consecutive wires for a device

mariaschuld opened the pull request:
**Context:**

This is the main PR of the wires management refactor. It refactors PennyLane to not assume any more that wires are consecutive integers that can be sorted and compared.

**Description of the Change:**

The main change is to use the ``wires`` argument in ``device('XXX', wires=..)`` _optionally_ to pass a sequence of representations or identifiers of wires, i.e. (to name a crazy example).

```
devlce('default.qubit', wires=['q1', 'q2', 0, 2, 1])
```

This user wire ordering is converted to a Wires object and stored in the new attribute ``register``. If the old version ``device('XXX', wires=N)`` is used, the register is initialised with  ``Wires(range(N))`` instead.

The user can now address wires with her labels, i.e. ``Hadamard(wires=['q2'])``. The device translates these labels into indices of the wires in the register, i.e. via

 ``wire_indices = self.register.indices(op.wires) ``

Finally, with this changes we can now let the Wires objects (which do not follow a consecutive logic) pass freely through PennyLane, instead of converting them to lists, as done in the previous PR. 

This PR also contains new tests to check that various parts of PennyLane work smoothly for non-consecutive wires.

josh146 commented on pennylane/operation.py:
```diff
@@ -1070,7 +1070,9 @@ def eigvals(self):
         # the eigenvalues is correct
-        obs_sorted = sorted(self.obs, key=lambda x: x.wires.tolist())
+        obs_sorted = sorted(
+            self.obs, key=lambda x: x.wires.tolist()
```
Good question 🤔 There must be a better way to extract the eigenvalues in wire order than the way it is implemented here.

I'm also still not sure how to deal with the case

```python
Hermitian(A1, wires=[0, 2]) @ Hermitian(A2, wires=[1, 3, 4])
```

I think it requires a rethink of the logic below?

mariaschuld replied:
I never really understood why one cannot use the order that the observables are saved in ``obs``...

josh146 replied:
Oh, so when the Tensor observable is created, just re-order the internal `self.obs` list according to wires? That could work

mariaschuld replied:
No, I thought just take the order that the user defined ... I guess that's not possible.

Could we sort them by hash of the observables? That's a unique order!

josh146 commented on pennylane/operation.py:
```diff
@@ -1259,7 +1262,9 @@ def loc(w):
-            for k, w in enumerate(self.wires.tolist()):
+            for k, w in enumerate(
+                self.wires.tolist()
+            ):  # TODO: refactor CV functions to not depend on consec wires
```
What else needs to change for CV operations?

mariaschuld replied:
the ``hermitian_expand`` method seems to depend on consecutive wire ordering, but it's extremely hard to understand...

josh146 commented on pennylane/plugins/default_qubit.py:
```diff
@@ -325,11 +328,18 @@ def reset(self):
+
+        Returns:
+            array: vector of marginal probabilities
+        """
```
:+1:

josh146 replied:
Oh actually this is inherited from QubitDevice, so doesn't need a docstring?

josh146 replied:
This is part of the QubitDevice API, so I'm not sure should change, actually? Since some devices might want qubit label rather than index? I'm not 100% sure.

mariaschuld replied:
That should be no problem. I could make all the ``qubit_device`` functions take ``wires`` in the QubitDevice API, and in this case just translate inside the function. Would that be more flexible?

josh146 replied:
I think so, yeah!

mariaschuld replied:
Sounds good, done.

mariaschuld replied:
I also rewrote the ``device`` in a similar manner, so it passes the wires to the devices, which then need to extract the ``registers`` or indices of the wires on the register...

(I hoped previously we could do this after the ``Device`` class refactor, but better to do it now!)

The internal devices are refactored in this PR, but we need to update all plugins when merging this PR. I'll start making PRs in the plugin repos.

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -359,18 +359,21 @@ def _remove_operator(self, operator):
-        for w in operator.wires.tolist():  # TODO: re-assess for for nonconsec wires
-            if w < 0 or w >= self.num_wires:
+        for w in operator.wires:
+            if w not in self.device.register:
```
much clearer!

josh146 commented on qchem/pennylane_qchem/qchem.py:
```diff
@@ -414,7 +414,8 @@ def _terms_to_qubit_operator(coeffs, ops):
 
         # wire ids
-        wires = op.wires.tolist()
+        wires = op.wires.tolist()  # Can we use subsystems here? Otherwise the Qchem library relies on users
+                                   # and devices assuming consecutive indices as wires
```
Would it be best to use wire objects/labels throughout the QChem library?

josh146 replied:
🤔

mariaschuld replied:
I agree. On my radar.

josh146 commented on tests/circuit_graph/test_circuit_graph.py:
```diff
@@ -211,9 +211,9 @@ def test_diagonalizing_gates(self):
         assert not circuit.is_sampled
 
-        circuit = CircuitGraph([qml.expval(qml.PauliX(0)), qml.sample(qml.PauliZ(1))], {})
+        circuit = CircuitGraph([qml.expval(qml.PauliX(0)), qml.sample(qml.PauliZ(1))], {}, Wires([0, 1]))
```
This addition must have been a pain to change everywhere!

mariaschuld replied:
There were worse pains along this PR ;)

josh146 commented on tests/test_device.py:
```diff
@@ -26,60 +26,80 @@
 
 
 @pytest.fixture(scope="function")
 def mock_device_with_observables(monkeypatch):
```
🔥

josh146 commented on tests/test_device.py:
```diff
@@ -136,75 +168,84 @@ def test_supports_operation_argument_types(self, mock_device_with_operations):
 
-        assert not mock_device_with_operations.supports_operation("S")
-        assert not mock_device_with_operations.supports_operation(qml.CNOT)
+        assert not dev.supports_operation("S")
```
This minor change makes the tests a lot easier to read.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -191,31 +211,35 @@ def execute(self, queue, observables, parameters={}, **kwargs):
 
                 if obs.return_type is Expectation:
-                    results.append(self.expval(obs.name, wires, obs.parameters))
+                    results.append(self.expval(obs.name, subsystems, obs.parameters))
```
Exactly. I would have hoped that in future the ``Device`` class could also just pass the operator to the apply function, that is so neat.

If a device inherits from ``Device`` and needs direct access to the wires they could always overwrite ``execute``.

Another option is to move the conversion to the existing devices inheriting from ``Device``

josh146 replied:
Perfect 👍 

We are planning to refactor `Device` to bring it inline with `QubitDevice`, and pass Operator objects rather than serialized data

mariaschuld replied:
Great, that will make things easier. Feel free to tag me on that task.

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -385,56 +385,56 @@ def analytic_probability(self, wires=None):
         raise NotImplementedError
 
-    def estimate_probability(self, wires=None):
+    def estimate_probability(self, subsystems=None):
```
my current favourite is ``wire_indices``, which is clear, but long!

mariaschuld replied:
We could also keep the name ``wires`` and just do the translation on the same variable name. But this would be extremely opaque to coders. The advantage is that we would not have to refactor each and every plugin that overwrites the functions...

mariaschuld replied:
I refactored everything with ``registers``.

It is not perfect, because the devices' ``register`` is an object like ``Wires=[0, 3, 'a']``, and ``registers`` in this register could also be interpreted as the wire labels. But somehow ``registers`` has a neutral connotation of ``the ith register``...

mariaschuld commented on pennylane/circuit_graph.py:
```diff
@@ -141,44 +140,49 @@ class CircuitGraph:
+        self.num_wires = len(register)
         """int: number of wires the circuit contains"""
         for k, op in enumerate(ops):
-            self.num_wires = max(self.num_wires, max(op.wires.tolist()) + 1)
```
Great :)

I think it's really nice to have access to the device's register everywhere in the qnode...

co9olguy commented:
Due to the feature freeze coming up, I'd recommend this one not get merged right away, even once approved

josh146 commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -177,27 +151,28 @@ def resolve_representation(self, grid, representation_grid):
-                )
+            for subsystem, operator in enumerate(grid.layer(i)):
+                wire = self.active_register[subsystem]
+                representation_layer[
+                    subsystem
+                ] = self.representation_resolver.element_representation(operator, wire)
```
Wow, this is pretty bad. I haven't noticed black doing this before

josh146 commented on tests/circuit_graph/test_qasm.py:
```diff
@@ -31,7 +31,7 @@ class TestToQasmUnitTests:
     def test_empty_circuit(self):
```
Good question. I had a look at the OpenQASM spec again; it allows you to name the register as a whole, but not the individual wires. Wires are always specified by consecutive indexing.

You _could_ create a separate register per wire, like so:
```
qreg q[1];
qreg r[1];

cnot q[0], r[0];
```

but this seems to be a very 'non-standard' way of using OpenQASM, and might run into issues with different frameworks.

Probably best to leave it as is for now.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -451,6 +475,17 @@ def check_validity(self, queue, observables):
+        Return:
+            list[int]: indices at which the wires are located on this device
+        """
+        return self._register.indices(wires)
```
That's useful, yes. I got this error quite a lot!

mariaschuld commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -347,12 +327,20 @@ def draw(self):
+            rendered_string += s.format(wire_names[i], 2 * self.charset.WIRE)
 
             for s in wire:
                 rendered_string += s
```
Yes, these changes are very minuscule...I think I understand what is going on though, and the tests are quite thorough.

mariaschuld commented on pennylane/templates/state_preparations/mottonen.py:
```diff
@@ -289,14 +289,14 @@ def MottonenStatePreparation(state_vector, wires):
     # code to avoid inverting at the end
 
     # Apply y rotations
-    for k in range(n_wires, 0, -1):  # Todo: use actual wire ordering!
```
I know, I am also quite surprised. But I need to test this properly.

mariaschuld replied:
Is tested now!

mariaschuld commented on pennylane/_device.py:
```diff
@@ -472,7 +498,7 @@ def expval(self, observable, wires, par):
         Args:
             observable (str or list[str]): name of the observable(s)
-            wires (List[int] or List[List[int]]): subsystems the observable(s) is to be measured on
+            wires (Wires): wires the observable(s) is to be measured on
```
Now expval, var and sample take ``Wires`` arguments, and the devices have to translate to registers...

mariaschuld commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -180,22 +182,22 @@ def _clear_network_data(self):
         self.mps = None
         self._contracted_state_node = None
 
-    def _add_node(self, A, wires, name="UnnamedNode", key="state"):
```
All private functions use ``registers``...

mariaschuld commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -80,7 +80,9 @@ class DefaultTensor(Device):
       >>> mps_tensornet = qml.device("default.tensor", wires=2, representation="mps")
```
This device needed to be refactored now that ``Device`` passes ``Wires`` in the ``execute`` function to ``apply``.

mariaschuld commented on pennylane/wires.py:
```diff
@@ -173,11 +173,14 @@ def indices(self, wires):
 
         Args:
-            wires (Wires or Iterable[Any]): Iterable containing the wires whose indices are to be found
+            wires (Wires or Iterable[Any] or Integral, str): Iterable containing the wires whose indices are to be found,
```
I replaced all types with a consistent format using ``Union[...]``. It's still lengthy, but at least exact.

mariaschuld commented on tests/circuit_drawer/test_circuit_drawer.py:
```diff
@@ -231,11 +237,17 @@ def test_resolve_decorations(self, dummy_circuit_drawer, grid, target_representa
+        raw_observable_grid = [[None] for _ in range(len(raw_operator_grid))]
+
+        drawer = CircuitDrawer(raw_operator_grid, raw_observable_grid,  Wires(range(10)))
+        drawer.move_multi_wire_gates(operator_grid)
```
Yes, the ``dummy_circuit_drawer`` fixture was created with some random grid, and then its method is tested with another grid called ``operator_grid``. This was a convenient cheat, but now the creation of the drawer remembers the register of all wires in the operations, and if they are not the same between the two grids there are problems.

These changes to the test use the ``operator_grid`` to construct the drawer in the first place, and use a dummy (``None``) observable grid.

mariaschuld commented:
@co9olguy Could you have another look at the refactor of ``default_tensor``?

Josh remarked that the ``Device`` should pass the original ``Wires`` to the ``apply``, ``probability`` etc. functions, so the subdevices have access to them. That required ``default_tensor`` to extract the ``registers`` in the right places...

co9olguy commented on tests/templates/test_layers.py:
```diff
@@ -451,7 +451,7 @@ def test_random_layer_randomwires(self, n_subsystems):
             )
 
-        wires = [q._wires for q in rec.queue]
+        wires = [q._wires.tolist() for q in rec.queue]
```
How come this one needs a `tolist()`, but it was removed in most other places?

mariaschuld replied:
The test computes the mean of randomly picked wires, so we need the integer representations here...

co9olguy commented on tests/circuit_drawer/test_circuit_drawer.py:
```diff
@@ -93,7 +93,7 @@ def to_layer(operation_list, num_wires):
 
     for op in operation_list:
-        for wire in op.wires:
+        for wire in op.wires.tolist():
```
how come `tolist()` is still needed here? :thinking:

mariaschuld replied:
I was just too lazy to rewrite the entire tests from bottom up - we can just avoid this function in tests that use the non-consec wire feature for now? Will add a todo though, would that be ok?

co9olguy commented on pennylane/_device.py:
```diff
@@ -206,9 +206,7 @@ def execute(self, queue, observables, parameters={}, **kwargs):
             self.pre_apply()
 
             for operation in queue:
-                # indices of the wires in the device's register
```
So now up to devices to implement the translation of wires to registers/subsystems?

mariaschuld replied:
Yep!

co9olguy commented on pennylane/_qubit_device.py:
```diff
@@ -378,65 +380,67 @@ def analytic_probability(self, wires=None):
+        else:
+            wires = Wires(wires)
+        # get indices of wires on the device's register
+        registers = self.wire_map(wires)
```
:ok_hand:

co9olguy commented on pennylane/operation.py:
```diff
@@ -474,7 +474,7 @@ def wires(self):
 
         Returns:
-            list[int]: wires
+            Wires: wires
```
Checks out :+1:

co9olguy commented on pennylane/plugins/default_qubit.py:
```diff
@@ -328,18 +329,17 @@ def reset(self):
         self._pre_rotated_state = self._state
 
-    def analytic_probability(self, subsystems=None):
-        """Return the (marginal) analytic probability of each computational basis state.
```
docstring unnecessary?

mariaschuld replied:
I learnt from @josh146 that functions inherited from a parent class do not need a docstring?

josh146 replied:
Yep!

co9olguy commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -337,34 +343,34 @@ def _add_state_prep_nodes(self, operation, par):
-                self.mps.apply_one_site_gate(op_node, wire)
-                self._free_wire_edges[wire] = self.mps.nodes[wire][1]
-            elif len(wires) == 2:
-                if abs(wires[1] - wires[0]) == 1:
+            if len(registers) == 1:
+                reg = registers[0]
+                self.mps.apply_one_site_gate(op_node, reg)
+                self._free_wire_edges[reg] = self.mps.nodes[reg][1]
+            elif len(registers) == 2:
+                if abs(registers[1] - registers[0]) == 1:
                     # TODO: set ``max_singular_values`` or ``max_truncation_error``
-                    self.mps.apply_two_site_gate(op_node, *wires)
-                    for wire in wires:
-                        self._free_wire_edges[wire] = self.mps.nodes[wire][1]
+                    self.mps.apply_two_site_gate(op_node, *registers)
+                    for reg in registers:
+                        self._free_wire_edges[reg] = self.mps.nodes[reg][1]
```
This is one place where I suspect things would break if `wire_map` did not return a consecutive sequence of indices, starting from zero. The reason for that is because `self.mps.nodes` is an object defined in the `tensornetwork` library, which implicitly indexes the nodes that way

mariaschuld replied:
I will add tests for this! But since the results of wire_map are indices, I would not see a user-invoked case where this can go  wrong, as long as the device is coded up correctly!

co9olguy commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -503,13 +522,13 @@ def _ev_exact(self, obs_nodes, wires):
-            meas_wires.extend(obs_wires)
-            for idx, w in enumerate(obs_wires):
+        for obs_node, obs_registers in zip(obs_nodes, registers):
+            meas_wires.extend(obs_registers)
+            for idx, w in enumerate(obs_registers):
                 # Use convention that the indices of a tensor are ordered like
                 # [output_idx1, output_idx2, ..., input_idx1, input_idx2, ...]
                 output_idx = idx
-                input_idx = len(obs_wires) + idx
+                input_idx = len(obs_registers) + idx
                 tn.connect(obs_node[input_idx], ket[w])  # A|psi>
                 tn.connect(bra[w], obs_node[output_idx])  # <psi|A
```
This is another place where things *might* break if `registers` is not a consecutive list, starting from 0. You can see there is some manual indexing going on, with the implicit assumption (in `enumerate`) that things start with 0

co9olguy commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -523,29 +542,29 @@ def _ev_exact(self, obs_nodes, wires):
             # but this would require change of `expval` behaviour and
             # refactor of `execute` logic from parent class
-            expval = self.mps.measure_local_operator(obs_nodes, wires[0])[0]
+            expval = self.mps.measure_local_operator(obs_nodes, registers[0])[0]
```
another place where things would likely break if `registers` is not a consecutive, zero-first list, since we are calling a function from an external library

mariaschuld commented:
> Hi @mariaschuld, I took a look over, and I think that the tensor network plugins, though they might pass, would not give the expected behaviour if `registers` is not of the form [0,1,2,...N-1]. I think we'll need some stronger tests to properly catch those bugs (the tests were written without this feature in mind)

Yes @co9olguy , the final PR will add tests everywhere for this feature! This may uncover some issues too. I will fork the final PR off this branch so that the version we merge is tested properly.

mariaschuld commented on pennylane/wires.py:
```diff
@@ -103,6 +103,22 @@ def __eq__(self, other):
             return self.wire_tuple == other.wire_tuple
         return False
 
```
Adding those so we can do `assert Wires(['q1, 2]) > Wires([1, 'a'])`, which is needed when we use ``groupby`` in various places to sort observable eigenvalues...

co9olguy replied:
See comment in related tests below

mariaschuld commented on pennylane/wires.py:
```diff
@@ -155,9 +171,12 @@ def index(self, wire):
-            return self.wire_tuple.index(wire.wire_tuple[0])
+            wire = wire.wire_tuple[0]
 
-        return self.wire_tuple.index(wire)
```
Better to return a legible error here, this came up in discussion with Nathan I think.

co9olguy commented on pennylane/plugins/default_gaussian.py:
```diff
@@ -852,30 +853,30 @@ def reset(self):
+        ind = np.concatenate([np.array(wire_indices), np.array(wire_indices) + self.num_wires])
         rows = ind.reshape(-1, 1)
         cols = ind.reshape(1, -1)
 
```
I think one of the hoped-for benefits of a wire management class is that both the user, and the plugin developer, would only have to worry about how the device works and not any bookkeeping. However, that doesn't really work in this particular plugin, since (correct me if I misunderstand) some functions need to know both about the "user" wires and the "device" register :thinking:

mariaschuld replied:
Yes, exactly. I will reply to your general comment about that.

mariaschuld replied:
Not really (if I understand your comment correctly), the plugin developer needs to know that the user creates a device with a register of custom labels, and that one can get the index of a subsystem using the `indices_on_register` function. Nothing needs to be known about the actual labels. 

A developer could also force the device to have a fixed register, in which case s/he has control over the labels as well, for example where a hardware comes with specific wire names.

mariaschuld replied:
Even though most devices just work on the `wire_indices`, this system makes sure that this object is very controlled, it is always a list of integers addressing subsystems that exist on the device.

mariaschuld replied:
In this case it is because the ``poly_quad_expectations`` is not a method of the CV operator class. So it cannot just get the attribute. It's more a helper function...

co9olguy commented on pennylane/operation.py:
```diff
@@ -1262,9 +1258,7 @@ def loc(w):
-            for k, w in enumerate(
-                self.wires.tolist()
-            ):  # TODO: refactor CV functions to not depend on consec wires
+            for k, w in enumerate(wire_indices):
```
It seems that this function (and related ones) worked before by accessing a class attribute. How come we can't just make `wire_indices` an attribute instead of passing it around from function to function?

mariaschuld replied:
I'm not sure what you mean exactly. `wire_indices` only gets passed from `heisenberg_obs` to `heisenberg_expand` (in the `CVObservable` class), or from `heisenberg_tr` to `heisenberg_expand` (in the `CVOperation` class). `heisenberg_tr` and `heisenberg_expand` get called by the qnode.

Of course, we could make those two functions set an attribute and then `heisenberg_expand` calls it. Is that what you mean? I always thought something like that is to be avoided, because it assumes that the attribute was always set before `heisenberg_expand` is called...But happy to do it that way!

co9olguy replied:
I was trying to understand why it was necessary to introduce a `wire_indices` arg to this function, when it wasn't needed before (from what I could determine [correct me if I'm wrong], this decision propagates to many other functions in `default.gaussian` then needing cosmetically updated signatures). I was wondering how we could avoid that. 

It seems before that this function accessed an attribute `self.wires.tolist()` rather than needing something to be passed. I was wondering if we could save all the redundant signature arguments that were introduced by using a similar trick here

mariaschuld replied:
Unfortunately not. This is really the one ugly side effect of the refactor, which only surfaced when I dug deep into the `heisenberg_expand` function. 

But let me try to motivate while it is rather a case of uncovering a structure that was there all along than just making devices unnecessarily complicated:

The qnode calls `heisenberg_expand` to "transform" an operator it encounters (through `heisenberg_tr` and `heisenberg_obs`). But the transformation depends on where a wire is located on the device, because that determines what dimensions in the matrix to change. In other words, the transformation needs information on the device, and these functions are therefore not "local" to the operator class (which means you could just create an operator and then use them), but they require a device to make sense.

The previous "locality" was in some sense an artefact based on the assumption that the labels the operator knows of are locations of a wire on the device. But we are removing that assumption. So the function now needs to get information about the device, namely the register.

co9olguy commented on qchem/pennylane_qchem/qchem/structure.py:
```diff
@@ -383,7 +383,7 @@ def _qubit_operator_to_terms(qubit_operator):
+                qml.operation.Tensor(*[xyz2pauli[q[1]](wires=q[0]) for q in term])  # TODO: nonconsecutive wires
                 if term
                 else qml.operation.Tensor(qml.Identity(0))
                 # example term: ((0,'X'), (2,'Z'), (3,'Y'))
```
yikes, just realized this is a 12-line list comprehension! :astonished:

mariaschuld replied:
Quite something. By the way, am working on qchem's wire management with Alain.

co9olguy commented on qchem/pennylane_qchem/qchem/structure.py:
```diff
@@ -383,7 +383,7 @@ def _qubit_operator_to_terms(qubit_operator):
             (
                 coef,
-                qml.operation.Tensor(*[xyz2pauli[q[1]](wires=q[0]) for q in term])
+                qml.operation.Tensor(*[xyz2pauli[q[1]](wires=q[0]) for q in term])  # TODO: nonconsecutive wires
```
When will this TODO be addressed? As I understand it, it will have buggy behaviour if the user does not have consecutive wires?

mariaschuld replied:
Yes, I am waiting for feedback on how quick we can refactor qchem, and if that's more than a few lines I will raise a warning for now, forcing qchem users to use consecutive wires.

co9olguy replied:
Sounds like a good plan :+1:

mariaschuld replied:
Will need to raise the warning, this could take a bit longer...

co9olguy commented on tests/test_device.py:
```diff
@@ -406,6 +406,20 @@ def test_args(self, mock_device):
         with pytest.raises(qml.DeviceError, match="specified number of shots needs to be at least 1"):
             Device(mock_device, shots=0)
 
+    def test_wire_map(self, mock_device):
```
should update name here, since the function no longer exists :crying_cat_face:

mariaschuld replied:
Nice one, thanks!

co9olguy commented on tests/circuit_drawer/test_circuit_drawer.py:
```diff
@@ -93,6 +93,7 @@ def to_layer(operation_list, num_wires):
     layer = [None] * num_wires
 
     for op in operation_list:
+        # TODO: Currently this function only works if the device's wires are consecutive integers
```
When will this be tackled?

mariaschuld replied:
It is low priority IMO, because we only use this function in tests with consecutive wires at the moment.

co9olguy commented on tests/circuit_drawer/test_circuit_drawer.py:
```diff
@@ -602,6 +602,38 @@ def drawn_qubit_circuit_with_probs():
+        + " q2: ─────├C──┤     \n"
+        + " -1: ──X──│───┤     \n"
+        + "  b: ──X──╰C──┤     \n"
+    )
```
Cool :sunglasses:

co9olguy commented on tests/templates/test_integration.py:
```diff
@@ -647,3 +647,82 @@ def circuit():
+            template(**nondiffable)
+            return qml.expval(qml.Identity(wires=0))
+
+        # construct qnode with nonconsecutive wires
```
Good to fix the seed explicitly in the file, and verify that that particular seed gives nonconsecutive wires (small probability it wouldn't). If the seed is not set explicitly, there's a chance that the behaviour of some (stochastic) tests will be dependent on what other tests are run

mariaschuld replied:
Good point actually! I decided to rather use a non-consecutive fixed base array from which the wires are sliced.

trbromley replied:
Has this change been pushed? I still see an `np.random.choice` below?

mariaschuld replied:
Sorry, not yet...refactoring this test is still a larger task!

co9olguy commented on tests/templates/test_integration.py:
```diff
@@ -647,3 +647,82 @@ def circuit():
+        if 'ph' in kwargs_nonconsec:
+             kwargs_nonconsec['ph'] = [nonconsecutive_wires[:3], nonconsecutive_wires[1:]]
+        if 'pphh' in kwargs_nonconsec:
+             kwargs_nonconsec['pphh'] = [[nonconsecutive_wires[:2], nonconsecutive_wires[2:]]]
```
This way of doing things is not very forward-compatible :crying_cat_face: (e.g., there could be future templates that satisfy these same conditions). Better to be explicit about testing different templates that require different conditions separately

co9olguy commented on tests/templates/test_integration.py:
```diff
@@ -647,3 +647,82 @@ def circuit():
+        circuit_consec()
+        circuit_nonconsec()
+
+        assert np.allclose(dev_consec._state[0], dev_nonconsec._state[0])
```
There's a lot going on in these tests that is masking what is actually being tested. Suggest simplifying, even if it means some code duplication

mariaschuld replied:
You are right - the tests are terrible to read. My hope was to not make new fixtures, because at the moment a template developer can just add a standard template to the list of fixtures and all tests are taken care of. But I know it is hacky! I will rewrite this, even though it may require the entire integration test structure to be rewritten :/

co9olguy commented on tests/test_wires.py:
```diff
@@ -110,6 +110,21 @@ def test_equality(self):
+        assert wires2 > wires3
+        assert wires2 > wires1
+        assert not wires1 < ['q1', 2, 1]
+        assert not wires2 > [1, 'q2', 3]
```
Unfortunately, I had no idea what to expect from `>` and `<`. :thinking:.

co9olguy replied:
A bigger issue is that Python orders sequences (like strings and lists of ints) in a way that I think is incompatible with what you want to do here.

Consider the following example (copied from my Python terminal):
```
>>> [0,1]<[1,2]
>>> True  # seems legit
>>> [0,1]<[1,-1]
>>> True  # ok, maybe something wrong with negative numbers?
>>> [0,1]<[1,0]
>>> True  # :thinking: 
>>> [0,1]<[1,"hi"]
>>> True  # no idea if this should be true or false
```
Note that this could lead to situations where `assert wires1 > wires2` and `assert wires2 > wires1` both hold

co9olguy replied:
https://stackoverflow.com/questions/30964157/python-inequality-operators-comparing-lists

mariaschuld replied:
This functionality would only be for the `groupby` function, so the goal is to have *any* comparison that is reproducible. But you are right, there might be side effects that would be non-intuitive. An alternative I thought of was to implement a comparison of string representations? Or to rewrite the entire eigenvalue sorting system in PL, which I'd rather avoid.

co9olguy replied:
Sadly, string comparison works the same as list comparisons.

The current implementation might provide self-consistent comparisons (i.e., the value of a comparison is unambiguously `True` or `False` every time), but it doesn't provide any self-consistent *ordering*. Which one of these do we need for eigenvalue sorting? (I'd assume the second)

trbromley replied:
I couldn't quite follow the above, is it possible to have a quick summary of why we want the ability to place an ordering?

josh146 replied:
I think we might have to revisit `Tensor.eigvals` and `Tensor.matrix` in an ADR, my gut feeling is that the current implementation should be rethought.

@trbromley I believe this context is this line, from the `Tensor.eigvals` property:

```python
        # Sorting the observables based on wires, so that the order of
        # the eigenvalues is correct
        obs_sorted = sorted(self.obs, key=lambda x: x.wires)
```

This sorting is to ensure that the returned eigenvalues are in the correct order for the tensor observable.

Essentially: we need the returned eigenvalues to be in the same order as they would be if we manually diagonalized the tensor observable in the computational basis (e.g., the returned eigenvalues must be in the same order as `diag(Lambda)`, where `Lambda = S.H.S^ct`, `H` is the observable, and `S` is some unitary change of basis operation that diagonalizes `H`).

Currently, we attempt to sort the observables in the tensor product by wire location, and then directly access the eigenvalues for each term in the product. This is to avoid doing any expensive diagonalization. But there are a lot of assumptions baked in here; it will work fine for
```python
qml.PauliZ(0) @ qml.Hermitian(A, wires=[2, 4]) @ qml.PauliX(1)
```
but (I'm pretty sure?) isn't well defined for something like

```python
qml.PauliZ(0) @ qml.Hermitian(A, wires=[2, 4]) @ qml.PauliX(3)
```

josh146 replied:
Thinking about this some more: we know that the diagonalization operation is given by

```python
def diagonalizing_gates(self):
    diag_gates = []
    for o in self.obs:
        diag_gates.extend(o.diagonalizing_gates())

    return diag_gates
```

Perhaps we can just modify this method so that we can keep track of the observable order, and make sure to use the _same_ order in `Tensor.eigvals`.

co9olguy replied:
I think I let myself get overly confused here. After further discussion with @mariaschuld, I think this method should give us a way to define a total order whenever we need it

co9olguy commented:
One more thing (this is as good a time as any): we should change the names of the class `ActsOn` to `WireEnums`

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -480,12 +480,12 @@ def marginal_prob(self, prob, wires=None):
+        wire_indices = self.indices_on_register(wires)
 
-        registers = np.hstack(registers)
+        wire_indices = np.hstack(wire_indices)
```
Actually, I don't get what the hstack is supposed to do here (but didn't want to introduce unecessary changes). Since `wires_indices` is just a list of integers, this converts the list into an array.  @josh146 would you know if this was catering for the cases of nested wires before?

josh146 replied:
It was, I believe! If you remove it and the tests pass, it _should_ be safe...

mariaschuld commented on pennylane/_device.py:
```diff
@@ -465,7 +465,7 @@ def check_validity(self, queue, observables):
                         )
                     )
 
-    def wire_map(self, wires):
```
The reason for a change is that a wire map sounds like it is a map from one labelling to another - but it is unclear what the new labelling actually is. What this function does is literally getting the wires' indices on the register, hence `indices_on_register`. If a developer does not know what a register is, s/he won't understand what this function does in the first place...

After long discussions with a few people about the names, I decided to go with "maximum explicit" here.

co9olguy replied:
ok so maybe time to admit my ignorance. "register" does not have an intuitive meaning for me :laughing:

mariaschuld replied:
Wait until you code up another device :P

mariaschuld commented on pennylane/plugins/default_gaussian.py:
```diff
@@ -492,7 +492,7 @@ def set_state(state, wire, mu, cov):
 
 
-def photon_number(mu, cov, wires, params, total_wires, hbar=2.0):
+def photon_number(mu, cov, wires, params, total_wires, register, hbar=2.0):
```
The entire `default_gaussian` plugin should be overhauled. The way it is done is that there is an operator map, and if an observable is called, PL maps it to a function and calls the function instead. 

Since different functions have different signatures, what was done here is to fake the signature of the longest function in all other functions. I just added another unused parameter, but there are many others.

mariaschuld replied:
Ah, that's the historical context! :) I rewrote the code so that I only have to pass the ``register`` to the one function that actually uses it.

mariaschuld commented on pennylane/plugins/default_gaussian.py:
```diff
@@ -599,14 +601,16 @@ def poly_quad_expectations(mu, cov, wires, params, total_wires, hbar=2.0):
 
 
-def fock_expectation(mu, cov, wires, params, total_wires, hbar=2.0):
+def fock_expectation(mu, cov, wires, params, total_wires, register, hbar=2.0):
```
Yep. I would really like to see a better way, but this is a `default_gaussian` refactor that is independent of this PR.

mariaschuld commented on pennylane/plugins/default_gaussian.py:
```diff
@@ -793,12 +795,12 @@ def expval(self, observable, wires, par):
+        mu, cov = self.reduced_state(wire_indices)
         _, var = self._observable_map[observable](
-            mu, cov, registers, par, hbar=self.hbar, total_wires=self.num_wires
+            mu, cov, wire_indices, par, self.num_wires, self.register, hbar=self.hbar
```
Yes. Again, I would propose not to meddle with the device and focus on this already gigantic PR :) Let's put a refactor on the roadmap!

co9olguy replied:
I'm going to disagree here. We're already meddled with the device by changing the signature of several functions unnecessarily. If it's unavoidable to have to have to make changes, I'd rather the changes were easier to follow/justify (e.g., having a single conditional for the one special case) rather than try to force changes confusingly to several functions in order to not "break the abstraction"

mariaschuld replied:
Convinced, I will implement the if statement!

mariaschuld commented on pennylane/plugins/default_qubit.py:
```diff
@@ -334,12 +333,5 @@ def analytic_probability(self, wires=None):
-        else:
-            wires = Wires(wires)
-            # get indices of wires on the device's register
-            registers = self.wire_map(wires)
```
The last master merge must have made this comment "outdated". But if I reconstruct correctly, your comment affects one of the functions in `default_qubit` that should just pass the wires object forward to the function it calls. It was a bug uncovered by adding the tests.

co9olguy replied:
Just to make sure I understand: was the uncovered bug fixed by removing these lines? Or was it fixed elsewhere in the PR?

mariaschuld replied:
It was fixed by removing these lines.

trbromley commented on pennylane/templates/state_preparations/mottonen.py:
```diff
@@ -124,7 +124,7 @@ def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
 
     control_indices = [
         int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % num_selections], 2)))
-        for i in range(num_selections)  # TODO: re-asses for nonconsecutive wires
```
Is there still a plan to do a follow up PR? Asking as I noticed this post was from a while ago.

mariaschuld replied:
Oh sorry, no that PR the comment is talking about was already merged into this branch. No worries :relieved:

co9olguy commented on pennylane/_device.py:
```diff
@@ -451,6 +465,17 @@ def check_validity(self, queue, observables):
                         )
                     )
 
+    def indices_on_register(self, wires):
```
`register` could be mistaken for a verb there? :thinking:

josh146 replied:
True 🤔

co9olguy commented on .github/CHANGELOG.md:
````diff
@@ -2,13 +2,35 @@
+  ```pycon
+  >>> dev = qml.device("default.qubit", wires=5)
+  ``` 
+  [(#666)](https://github.com/XanaduAI/pennylane/pull/666)
````
Don't worry too much about the changelog, we always clean it up and rearrange before release

trbromley commented on tests/beta/test_default_tensor.py:
```diff
@@ -1754,3 +1754,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, rep, tol):
+                                                (['a', 'c'], [3, 0]),
+                                                ([-1, -2], ['ancilla', 2]),
+                                                (['a'], ['nothing']),
+                                                ])
```
No worries if this is a massive overhead now, but would it also make sense to have the "vanilla" case as well, e.g., `[0, 1, 2]`?

co9olguy replied:
I think that's a one-line change :grinning:

mariaschuld replied:
Added!

trbromley commented on tests/ops/test_qubit_ops.py:
```diff
@@ -1057,5 +1057,5 @@ def test_decomposition(self):
         decomp = qml.DiagonalQubitUnitary.decomposition(D, [0, 1, 2])
 
         assert decomp[0].name == "QubitUnitary"
-        assert decomp[0].wires == [0, 1, 2]  #Wires([0, 1, 2])
```
I admire your patience on this PR :smile:

mariaschuld replied:
I admire your patience making it up to here in the review! ;)

trbromley commented on tests/plugins/test_default_qubit_tf.py:
```diff
@@ -1236,3 +1237,55 @@ def cost(weights):
+        circuit1 = self.make_circuit_expval(wires1)
+        circuit2 = self.make_circuit_expval(wires2)
+
+        assert np.allclose(circuit1(), circuit2(), tol)
```
Is it too late to think of a way of having a common place for these tests?

mariaschuld replied:
Well, we kind of have...the shared device test suite in ``plugins/tests``. But it does not get run when you ``make test``, so I thought it's better to put the integration tests into every device for now.

@josh146, do you remember if we were planning to run the shared device tests automatically at some stage? They have a lot of flaky usage in them though!

trbromley commented on tests/test_qubit_device.py:
```diff
@@ -123,90 +147,96 @@ def test_op_queue_accessed_outside_execution_context(self, mock_qubit_device):
 
         with pytest.raises(
-            ValueError, match="Cannot access the operation queue outside of the execution context!"
+                ValueError, match="Cannot access the operation queue outside of the execution context!"
```
Was this a result of running `black`?

mariaschuld replied:
This is so strange! Must be...let me change it back!

trbromley commented on tests/test_qubit_device.py:
```diff
@@ -21,7 +21,6 @@
 import pennylane as qml
 from pennylane import QubitDevice, DeviceError
 from pennylane.qnodes import QuantumFunctionError
-from pennylane import expval, var, sample
```
Just for my understanding, how did these end up being not needed? Or were they extras also before?

mariaschuld replied:
If I remember correctly, they were obsolete before this PR...I sometimes get tempted to clean code, although it is not nice for code review.

josh146 commented on pennylane/vqe/vqe.py:
```diff
@@ -114,7 +114,7 @@ def __str__(self):
                 obs_strs = ["{}{}".format(OBS_MAP[i.name], i.wires[0]) for i in obs.obs]
```
```suggestion
                obs_strs = ["{}{}".format(OBS_MAP[i.name], i.wires.tolist()[0]) for i in obs.obs]
```

josh146 replied:
This is so the Hamiltonian continues to print correctly for Tensor observables.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -41,8 +43,9 @@ class Device(abc.ABC):
     Args:
-        wires (int): number of subsystems in the quantum state represented by the device.
-            Default 1 if not specified.
+        wires (Union[int, Iterable[Union[Number, str]]]): Number of subsystems represented by the device,
```
Good question. Technically yes, the user could provide ``Wires``, but it should be an internal class.

And exactly right, ``Number`` is used for those ``1.0`` inputs to pass. If the value is ``4.5`` everythign will still work fine, but the user has to address the wire by ``4.5`` all the time. The only condition for wire labels is that they are unique, which is checked automatically.

trbromley replied:
Oh I like it :smile: makes me want to use e.g. pi as a wire

mariaschuld commented on pennylane/_device.py:
```diff
@@ -41,8 +43,9 @@ class Device(abc.ABC):
-            Default 1 if not specified.
+        wires (Union[int, Iterable[Union[Number, str]]]): Number of subsystems represented by the device,
+            or iterable that contains unique labels for the subsystems as numbers (i.e., ``[-1, 0, 2]``)
+            or strings (``['ancilla', 'q1', 'q2']``). Default 1 if not specified.
```
I think maybe in the tests?

mariaschuld commented on pennylane/_device.py:
```diff
@@ -41,8 +43,9 @@ class Device(abc.ABC):
     Args:
-        wires (int): number of subsystems in the quantum state represented by the device.
-            Default 1 if not specified.
+        wires (Union[int, Iterable[Union[Number, str]]]): Number of subsystems represented by the device,
```
Sure! I thought the Union notation is good for automatic type hinting, but haven't investigated this.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -127,6 +136,12 @@ def shots(self):
         return self._shots
 
+    @property
+    def register(self):
```
Both would be good, but register is a special wires item: It is the sequence of *all* wires present/addressable in the computation.

mariaschuld replied:
I added an explanation in the code, where ``self._register`` is defined.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -451,6 +465,17 @@ def check_validity(self, queue, observables):
 
+    def indices_on_register(self, wires):
+        """Get the indices of the wires in the register of this device.
+
```
Great idea!

mariaschuld commented on pennylane/_device.py:
```diff
@@ -451,6 +465,17 @@ def check_validity(self, queue, observables):
+        """Get the indices of the wires in the register of this device.
+
+        Args:
+            wires (Wires): wires object
```
Very good point. I updated example and docstring.

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -216,8 +219,8 @@ def apply(self, operations, **kwargs):
-        >>> op.wires # wires that the operation acts on
-        [0]
+        >>> op.wires # returns a Wires object representing the wires that the operation acts on
+        Wires([0])
```
Well-spotted, thanks

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -391,20 +393,25 @@ def estimate_probability(self, wires=None):
+        else:
+            wires = Wires(wires)
+        # get indices of wires on the device's register
+        wire_indices = self.indices_on_register(wires)
```
Happy with that suggestion!

mariaschuld commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -180,22 +182,22 @@ def _clear_network_data(self):
         self._contracted_state_node = None
 
-    def _add_node(self, A, wires, name="UnnamedNode", key="state"):
+    def _add_node(self, A, wire_indices, name="UnnamedNode", key="state"):
```
Yes, I want to now stick to the convention that ``wires`` uses the user's wire labels, while ``wire_indices`` is a list of integer-valued indices.

mariaschuld commented on pennylane/beta/plugins/default_tensor.py:
```diff
@@ -209,7 +211,7 @@ def _add_node(self, A, wires, name="UnnamedNode", key="state"):
         return node
 
-    def _add_initial_state_nodes(self, tensors, wires, names):
+    def _add_initial_state_nodes(self, tensors, sequence_of_wire_indices, names):
```
Oh it is a sequence of lists of integers, like ([0, 1], [2, 4, 3],...). The name is really a bit long, would you have a better idea?

mariaschuld commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -45,6 +46,7 @@ class CircuitDrawer:
     Args:
         raw_operation_grid (list[list[~.Operation]]): The CircuitGraph's operations
         raw_observable_grid (list[list[qml.operation.Observable]]): The CircuitGraph's observables
+        register (Wires): register of all wires in the circuit
```
No, the user should only have to call ``myqnode.draw()``...The circuit drawer instance is then created in the qnode and gets most of its information from the circuit graph....

mariaschuld commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -117,55 +121,26 @@ def __init__(
-        return self._internal_wire_to_circuit_wire[wires]
+        all_wires_with_duplicates = [op.wires for op in all_operators if op is not None]
+        # make Wires object containing all used wires
+        all_wires = Wires.all_wires(all_wires_with_duplicates)
+        # shared wires will observe the ordering of the register
+        shared_wires = Wires.shared_wires([self.register, all_wires])
```
A sharp observation :) In this case I am using ``shared_wires`` because the function orders the resulting object with respect to the ordering in the first argument of the list (here ``self.register``). This way the circuit is drawn in the way that the register is structured. But you are right, ``shared_wires`` and ``all_wires`` contain the same wires, only perhaps in a different order.

mariaschuld commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -177,27 +152,28 @@ def resolve_representation(self, grid, representation_grid):
-                )
+            for wire_indices, operator in enumerate(grid.layer(i)):
+                wire = self.active_register[wire_indices]
+                representation_layer[
+                    wire_indices
+                ] = self.representation_resolver.element_representation(operator, wire)
```
I know, but that was "black" doing its magic...

mariaschuld commented on pennylane/operation.py:
```diff
@@ -365,8 +365,7 @@ def __init__(self, *params, wires=None, do_queue=True):
 
-        wires = Wires(wires)
-        self._wires = wires.tolist()  #: list[int]: wires on which the operator acts
+        self._wires = Wires(wires)  #: Wires: wires on which the operator acts
```
The property 
```
@property
    def wires(self):
```
from further below makes sure that ``self.wires`` actually calls this property function, which returns ``self._wires``. I also never understood why one does not just call the attribute ``self.wires`` and get it directly, but I assume using properties is somewhat safer.

mariaschuld commented on pennylane/plugins/default_qubit.py:
```diff
@@ -325,10 +329,9 @@ def reset(self):
 
     def analytic_probability(self, wires=None):
-        """Return the (marginal) analytic probability of each computational basis state."""
+
```
Yes, exactly. It seems that functions that overwrite parent functions don't need to redefine the docstring.

mariaschuld commented on qchem/pennylane_qchem/qchem/structure.py:
```diff
@@ -414,7 +414,8 @@ def _terms_to_qubit_operator(coeffs, ops):
 
         # wire ids
-        wires = op.wires
+        wires = op.wires.tolist()  # Can we use subsystems here? Otherwise the Qchem library relies on users
+                                   # and devices assuming consecutive indices as wires
```
Will discuss this today with the qchem team, but some functions rely on the fact that wires were assumed to be index locations on the register. We had the same issue for the qchem templates, which did things like 
```
for w in range(wires[0], wires[2]):
    qml.Hadamard(w)
```

But we don't even know if ``w`` exists on the device...

mariaschuld replied:
So it needs a restructure, but I think it is a good one!

zeyueN replied:
Hey @mariaschuld, what does "subsystems" mean here in the comment?

mariaschuld replied:
Oh I'm sorry @zeyueN , this is an anachronism from the past. I meant the wire indices. A Qnode and a quantum device always know which wire label refers to a qubit at which index (i.e. 'ancilla' -> 0, '0' -> 1, '1' -> 2...). 

But I think the comment does not make a lot of sense, unless functions of qchem are called from a qnode. 

Let me know if you need more info!

mariaschuld commented:
Regarding @josh146 [comment](https://github.com/XanaduAI/pennylane/pull/666#discussion_r458554278) on Tensor observables: I just realised that another problem here is the omission of identities, like in ``qml.PauliZ(0) @ qml.PauliX(2)``. Here the ``Tensor`` class "knows" that the second position should be ``qml.PauliZ(1)``. But what if  we have ``qml.PauliZ('q1') @ qml.PauliX('ancilla')``? The only place where PennyLane can interpret this is inside the qnode, because it knows the register...

It is another example of where the wires refactor uncovers how much the independence of the ``operator`` module relied on this implicit assumption that we use consecutive wire labels.

mariaschuld commented:
@josh146, @trbromley , @co9olguy  Thanks for your great reviews! No need for action for now, I will ping you when everything is ready (in case you want to have a last look).

As a summary, I addressed the major issues as follows:
- eigenvalue sorting: I reverted the additions to the wires class to compare ``Wires`` objects, and for now compare ``wires.tolist()`` objects in the ``groupby`` functions of the eigenvalue sorting. Working on solving this issue on a separate branch, and it is actually quite non-trivial.
- templates integration test being ugly: instead of refactoring that entire test module (which has many implications in how templates are added) I just significantly tidied up the tests and now check for the template's name when making special case exceptions (like setting the wires argument for UCCSD)
- qchem: Zeyue will kindly work on making qchem wire-order independent asap since it is not clear how to raise a user warning
- heisenberg_expand function: I very much tidied up here as well, most functions now take only the register and infer the rest. I think we can leave it like this until the CV devices get polished again
- ``ActsOn`` is renamed to ``WiresEnum``

zeyueN commented:
Hey @mariaschuld ! Just gonna put some of my observations/questions here about the wires, based on what I saw from within the context of qchem. Hopefully not hijacking the comments here and sorry for the long rant.:laughing:

The more I look into it the more it looks like there are multiple things at play here, definitely not as innocent as it initially seemed 😆 . To me, there are two assumptions under the old way of naming wires: 1) consecutive order, and 2) `Operator.wires` containing wires that match the wires in the now `Device.registry`. Breaking (1) while keeping (2) unbroken will introduce the side effect of making `Operator`s not independent of devices (ie an `Operator` has to know about the wire naming/arrangement conversions used by a device when it's defined BEFORE it's stuck into to a `QNode`, essentially making it weakly attached to certain devices from the outset). 

But to keep `Operator`s independent of devices, I feel assumption (2) has to be broken. This means `Operator.wires` needs to be conceptually different from `Device.registry`, which kind of makes sense based on my limited knowledge in Pennylane's usage. The former could be for users to name their own wires for whatever reason, and the latter is used for conforming to a devices wire arrangement. And this will require a mapping between the 2 set of wires, likely in qnode when appending operators where it currently raises the "invalid wire" error. Haven't put too much thought into how to do this mapping though (user supplied? naive guessing? some kind of device-specific compiling that takes devices limitations into considerations?) One hacky way, though, of keeping both (2) and device-independence untouched is to enforce users to use the qnode decorator whenever they want to use custom wiring, ie custom wired operators must be device dependent by being in a qnode.

In the context of qchem, if we keep (2) unbroken, the conversion from OpenFermion (who always uses consecutive int cubits) to Pennylane will require the user to supply the wire names of the device he latter is going to run the ansatz on; this could even be before a device is defined. This feels a bit unnatural to me and thus led to the ranting above. Or, I could keep wires as is and experiment with the operator-to-device wire mapping mentioned above.

And the issue about `Tensor` which we also use extensively in qchem (with the assumption of implicit `Identity`). My gut feeling is that, if the `Tensor`'s wires doesn't have to conform to a device's registry, then there could be more freedom in telling it the order and all the implicit wires.

Also @agran2018, to keep everyone in the loop.

josh146 commented:
Thanks @zeyueN, some good thoughts. From a very high level, it would be great it we could preserve

* Operations remaining independent of the device

It would also be quite nice to preserve

* Assuming consecutive integers by default 

as this is a familiar paradigm in other QC libraries (Qiskit, OpenFermion), and allows us to abstract a lot of the more complicated logic. Users running simulations likely will be working with assumed consecutive integers as well, and this would allow them to avoid any additional 'wire label registration'. However, this one is likely highly non-trivial as @mariaschuld was mentioning.

co9olguy commented:
Thanks @zeyueN, your comment raises some good food for thought!

We're thinking of ways for make operators more independently composable outside of QNodes in the future, and this would require a strong separation from the specific devices. 

Conceptually, I think we can indeed separate "wire labels" logically from devices.
For instance, being able to do something like 
`qml.CNOT(wires=["alice", "bob"])`
is valuable to users, without ever attaching it to any implementation on a device.

co9olguy commented:
> 2) Operator.wires containing wires that match the wires in the now Device.registry
Is this the case? All `Operator.wires` needs to know about is the user's name for the wires, not how the device handles them, no? (can be assumed consecutive integers on device side, except for Rigetti, where they are hard-wired)

zeyueN commented:
> > 1. Operator.wires containing wires that match the wires in the now Device.registry
> 
> Is this the case? ideally, all `Operator.wires` needs to know about is the user's name for the wires, not how the device handles them, no? (can be assumed consecutive integers on device side, except for Rigetti, where they are hard-wired)

This is where I was referring to that matches operator's wires to device's wires:
https://github.com/XanaduAI/pennylane/blob/6a4828d2315555b5e3ecee810e94e5e29368d3e6/pennylane/qnodes/base.py#L370

co9olguy commented:
Thanks @zeyueN for finding that one, it definitely looks like something that should be ironed out.
Inspecting the code, it seems operators are weakly coupled to the device, but only if they are created inside a QNode's quantum function (e.g., `def circuit()`). 

I think what should be checked is that the user's labels for wires in the operations are self-consistent, and can be mapped to corresponding labels on the underlying devices, not necessarily that the user's names for wires matche the underlying labels the device prefers (by default, consecutive ordering)
